### ISSUE TITLE
string converted into number parsing numbers

### ISSUE THREAD

**Issue by imrafaelmerino** [ngs-doo/dsl-json#153]
Hi,

doing some testing, a special case called my attention. The methods to deserialize into numbers from the class NumberConverter, don't return an error when the number is a string, I mean:

{"a" : "1"} 

NumberConverter.deserializeInt() 

would parse and convert "1" into 1. From my point of view, it should return an error, or at least, it would be nice to pass over that decision to the client, because it's not clear that it is the expected behavior.

The case mentioned above can be even worse in the following scenario
{"a": "0001"}
because all the leading zeros are dropped, returning the integer 1

Any ideas or suggestions? 

Thanks in advance

**Comment by plokhotnyuk** [comment]
Leading zeros in a string are safe and most parsers support them, as example for Base-16 and Base-64 representations.

**Comment by imrafaelmerino** [comment]
I don't understand your answer. I think I didn't express correctly my question.
Taking a look at the **deserializeInt** method, you can see how double quotes are parsed, and then, any leading zero will be dropped after converting the string into an integer:

```
	public static int deserializeInt(final JsonReader reader) throws IOException {
		if (reader.last() == '"') {   // I think it shoud return an error, a string is not a number
			final int position = reader.getCurrentIndex();
			final char[] buf = reader.readSimpleQuote();
			try {
				return parseNumberGeneric(buf, reader.getCurrentIndex() - position - 1, reader).intValueExact();
			} catch (ArithmeticException ignore) {
				throw reader.newParseErrorAt("Integer overflow detected", reader.getCurrentIndex() - position);
			}
		}
		final int start = reader.scanNumber();
		final int end = reader.getCurrentIndex();
		final byte[] buf = reader.buffer;
		final byte ch = buf[start];
		if (ch == '-') {
			return parseNegativeInt(buf, reader, start, end);
		}
		return parsePositiveInt(buf, reader, start, end, 0);
	}

```

**Comment by plokhotnyuk** [comment]
A leading 0 indicates an octal number in JavaScript. So, it was declared as illegal in JSON specification for numbers, but for strings it is quite acceptable to have any number of zeroes in the beginning.

Jsoniter-scala also [allows leading zeros for stringified numbers](https://github.com/plokhotnyuk/jsoniter-scala/blob/0636c85ec70e0ff5f679bb5bf5118216f15f9f76/jsoniter-scala-core/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonReader.scala#L1179)

**Comment by zapov** [comment]
We can have configuration option if it should fail on such quotes which would default to not failing.

**Comment by imrafaelmerino** [comment]
I agree with you, but the method I'm talking about is **deserializeInt** and not **deserializeString**,
if I'd declare the key _a_ as an integer, I would expect that deserializing the string {"a": "001"} would return an error (the key a is not an integer). Taking into account the implementation of **deserializeInt**, the returned Json would be {"a" : 1}

**Comment by plokhotnyuk** [comment]
@plokhotnyuk yeap, the naming like `deserializeAsInt`  would be better for this routine.

**Comment by imrafaelmerino** [comment]
I'm using those methods, and I have two options:
-reimplementing them
-if we agree on a fix (a configuration option like @zapov  pointed out), implement it and make a pull request
What do you think? 

Thanks for your time

**Comment by zapov** [comment]
Configuration is fine as long as it affects all numbers behavior (built in ones at least).
Something like `allowNumbersInQuotes` or `allowNumbersAsString`
But I don't really find this valuable too much, because to me JSON is just a transport medium. It does not dictate the types at the end of the spectrum. Thus it's perfectly fine to encode numbers in quotes.
And technically if you don't do that for non-trivial numbers (eg, BigDecimals more precise than doubles, or BigInts longer than longs) you'll have hard time dealing with them as JSON only anyway.

**Comment by zapov** [comment]
Also, before you write any patch, please consider how this will affect NaNs and Infinities, as they cannot be expressed in JSON and thus require quoting by default.

**Comment by imrafaelmerino** [comment]
@Zapov I disagree, in the end, the consumer and producer has to agree on a schema and strings are not numbers. Assuming that "00001" is equal to 1 can be totally wrong in a lot of scenarios. 

I don't understand the second paragraph:
"And technically if you don't do that for non-trivial numbers (eg, BigDecimals more precise than doubles, or BigInts longer than longs) you'll have hard time dealing with them as JSON only anyway"

Thanks

**Comment by zapov** [comment]
I was looking into something unrelated and will make it like you suggested in the new version.
Leading 0 will not be allowed regardless if it's quoted or not.

My other comment was more about that NaN and Infinity cannot be represented as numbers in JSON and thus you always have to quote them.
The other common pattern is to quote numbers which are losing preceision when converted into Javascript, such as large BigDecimals.
This its not that uncommon to send quoted numbers around.

**Comment by zapov** [comment]
This specific issue should be resolved in v1.9.4. If there is any other, please open up a new Issue
